TACC-Cloud/hazmapper
# task/WG-74: Create Map React Modal

## Overview: ##
Create a new map modal in react.  See the current angular "Create new map" modal for comparison.

skip doing the file/project browser as it requires to much work.

Add POST request to Geoapi's /projectsendpoint by adding/using a new hook in react/src/hooks (and due to being blocked by WG-208,  it defaults to always creating a map linked with the users My Data.)

## PR Status: ##


## Related Jira tickets: ##

* [WG-74](https://tacc-main.atlassian.net/browse/WG-74)

## Summary of Changes: ##
- Created a new UI for creating a map project (custom modal)
- Added form validation with UI/UX for user (see screenshots)
- Added working POST request that creates new projects (currently using v2 endpoint, but Nathan and I are working on WG-212 to improve backend and get it ready to use v3) 
- Added unit tests for new CreateMapModal component
- system_path is now dynamic based off the currently logged in and authenticated user. This is used on the new modal for Save Location. When we add a File Browser System, this will need to change to include selected files.
- Used tacc's core-styles to update new modal's style and layout. In WG-246 we will update this to use FieldWrapperFormik from Tup's core-wrappers lib. 
- Added a useCreateProject hook to handle the POST request with the form data. 
## Testing Steps: ## 
1. Start up React frontend (npm ci and npm run dev)
2. Go to localhost:4200 in browser
3. On the Main Menu click the new "Create Map" button
4. Once Modal is open click the Submit button (without any input in the fields to verify form validation is working *should see red text prompting user to fill in missing input*)
5. Fill in the three fields (Name, Description, Custom File Name) with test input
6. Try using symbols other than a hyphen and/or underscore in the Custom File Name field and make sure form validation prompts user with correct message (see screenshot)
7. Once all three fields are populated, click the "Create" button. 
8. You should be taken to the map page (this map is just a placeholder from a set fixture, for now but the project uuid in the URL is correct)
9. Open dev tools in Chrome and check the console to see the data that was sent in POST request (if you decided to check the 'Sync Folder' box, you should see both observable and watch_content as true in the payload)
10. Go to Hazmapper prod and check that new project is present in project listing
11. You should also see that your "You have ___ projects" on the Main Menu for react has increased in number by one.

## UI Photos:
<img width="698" alt="Screenshot 2024-01-12 at 11 11 47 AM" src="https://github.com/TACC-Cloud/hazmapper/assets/50084480/7b674fa7-e046-4b43-8d8e-bdecca7e9490">
<img width="1179" alt="Screenshot 2024-01-12 at 11 12 06 AM" src="https://github.com/TACC-Cloud/hazmapper/assets/50084480/6ae308a6-759e-4591-81b6-f072660ba3f4">
<img width="741" alt="Screenshot 2024-01-12 at 11 12 18 AM" src="https://github.com/TACC-Cloud/hazmapper/assets/50084480/9a2c1c40-517d-44c2-9b0f-7aa1b86b91ff">
<img width="741" alt="Screenshot 2024-01-12 at 11 12 43 AM" src="https://github.com/TACC-Cloud/hazmapper/assets/50084480/db281145-e32d-494c-a899-08c845083f3c">

## Core Styles Added:
<img width="706" alt="Screenshot 2024-01-25 at 9 40 46 AM" src="https://github.com/TACC-Cloud/hazmapper/assets/50084480/c7f04447-62b6-4280-9b60-d3796da0d444">


## Notes: ##
This new Modal will require follow-up work relating to [WG-246](https://tacc-main.atlassian.net/browse/WG-246) (using FieldWrapperFormik), as well as adding the feature/behavior where the filename is updated by the map name (unless the user manually changes the filename after). See https://tacc-main.atlassian.net/browse/WG-249.
